crates/core/src/db/update.rs

```diff
@@ -1,627 +1,204 @@
+    // TODO: consider using `ErrorStream` here.
+    let old_def = plan.old_def();
+    for table in existing_tables
+        .iter()
+        .filter(|table| table.table_type != StTableType::System)
+    {
+        let old_def = old_def
+            .table(&table.table_name[..])
+            .ok_or_else(|| anyhow::anyhow!("table {} not found in old_def", table.table_name))?;
 
-        SchemaUpdates::Tainted(tainted) => {
-            system_logger.error("Module update rejected due to schema mismatch");
-            let mut tables = Vec::with_capacity(tainted.len());
-            for t in tainted {
-                system_logger.warn(&format!("{}: {}", t.table_name, t.reason));
-                if let TaintReason::IncompatibleSchema { existing, proposed } = t.reason {
-                    let existing = format!("{existing:#?}");
-                    let proposed = format!("{proposed:#?}");
-                    let diff = TextDiff::configure()
-                        .timeout(Duration::from_millis(200))
-                        .algorithm(Algorithm::Patience)
-                        .diff_lines(&existing, &proposed);
-                    system_logger.warn(&format!(
-                        "{}: Diff existing vs. proposed:\n{}",
-                        t.table_name,
-                        diff.unified_diff()
-                    ));
-                }
-                tables.push(t.table_name);
-            }
-            return Ok(Err(UpdateDatabaseError::IncompatibleSchema { tables }));
-        }
+        table.check_compatible(old_def)?;
     }
 
-    Ok(Ok(()))
-}
-
-/// The reasons a table can become [`Tainted`].
-#[derive(Debug, Eq, PartialEq)]
-pub enum TaintReason {
-    /// The (row) schema changed, and we don't know how to go from A to B.
-    IncompatibleSchema {
-        existing: TableSchema,
-        proposed: TableSchema,
-    },
-    /// The table is no longer present in the new schema.
-    Orphaned,
-}
-
-impl fmt::Display for TaintReason {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.write_str(match self {
-            Self::IncompatibleSchema { .. } => "incompatible schema",
-            Self::Orphaned => "orphaned",
-        })
+    match plan {
+        MigratePlan::Manual(plan) => manual_migrate_database(stdb, tx, plan, system_logger, existing_tables),
+        MigratePlan::Auto(plan) => auto_migrate_database(stdb, tx, plan, system_logger, existing_tables),
     }
 }
 
-/// A table with name `table_name` marked tainted for reason [`TaintReason`].
-#[derive(Debug, PartialEq)]
-pub struct Tainted {
-    pub table_name: Box<str>,
-    pub reason: TaintReason,
+/// Manually migrate a database.
+fn manual_migrate_database(
+    _stdb: &RelationalDB,
+    _tx: &mut MutTxId,
+    _plan: ManualMigratePlan,
+    _system_logger: &SystemLogger,
+    _existing_tables: Vec<Arc<TableSchema>>,
+) -> anyhow::Result<()> {
+    unimplemented!("Manual database migrations are not yet implemented")
```

Comment:
TODO: need to check for a reducer marked `LifeCycle::Update` and call it here? Or should it be called after auto-migration?

Reply:
We're dropping it (https://github.com/clockworklabs/SpacetimeDB/issues/1370), so this is fine.